api/types/headlessauthn.go

```diff
@@ -87,3 +87,86 @@ func (s HeadlessAuthenticationState) IsUnspecified() bool {
 func (s HeadlessAuthenticationState) IsPending() bool {
 	return s == HeadlessAuthenticationState_HEADLESS_AUTHENTICATION_STATE_PENDING
 }
+
+// headlessStateVariants allows iteration of the expected variants
+// of HeadlessAuthenticationState.
+var headlessStateVariants = [4]HeadlessAuthenticationState{
+	HeadlessAuthenticationState_HEADLESS_AUTHENTICATION_STATE_UNSPECIFIED,
+	HeadlessAuthenticationState_HEADLESS_AUTHENTICATION_STATE_PENDING,
+	HeadlessAuthenticationState_HEADLESS_AUTHENTICATION_STATE_DENIED,
+	HeadlessAuthenticationState_HEADLESS_AUTHENTICATION_STATE_APPROVED,
+}
+
+// Parse attempts to interpret a value as a string representation
+// of a HeadlessAuthenticationState.
+func (s *HeadlessAuthenticationState) Parse(val string) error {
+	for _, state := range headlessStateVariants {
+		if state.Stringify() == val {
+			*s = state
+			return nil
+		}
+	}
+	return trace.BadParameter("unknown request state: %q", val)
+}
+
+// HeadlessAuthenticationFilter encodes filter params for headless authentications.
+type HeadlessAuthenticationFilter struct {
+	Name     string
+	Username string
+	State    HeadlessAuthenticationState
+}
+
+// key values for map encoding of headless authn filter.
+const (
+	headlessFilterKeyName     = "name"
+	headlessFilterKeyUsername = "username"
+	headlessFilterKeyState    = "state"
+)
+
+// IntoMap copies HeadlessAuthenticationFilter values into a map.
+func (f *HeadlessAuthenticationFilter) IntoMap() map[string]string {
+	m := make(map[string]string)
+	if f.Name != "" {
+		m[headlessFilterKeyName] = f.Name
+	}
+	if f.Username != "" {
+		m[headlessFilterKeyUsername] = f.Username
+	}
+	if !f.State.IsUnspecified() {
+		m[headlessFilterKeyState] = f.State.Stringify()
```

Comment:
Why not use the protobuf string encoding here? We could check the `HeadlessAuthenticationState_value` map instead of relying on some ad-hoc encoding. We don't expect to be using this filter outside of the auth, but still.